File: day23/day23.py
#!/usr/bin/env python3

import heapq
import sys

energy = {
    "A": 1,
    "B": 10,
    "C": 100,
    "D": 1000,
}

target = {
    "A": 3,
    "B": 5,
    "C": 7,
    "D": 9,
}

class Map(object):
    def __init__(self):
        self.data = None
        self.energy = 0

    def from_txt(self, txt):
        self.data = tuple(x for x in txt.splitlines())

    def swap(self, a, b):
        x0, y0 = a
        x1, y1 = b

        data = list(self.data)
        t = data[y0][x0]
        data[y0] = data[y0][:x0] + data[y1][x1] + data[y0][x0+1:]
        data[y1] = data[y1][:x1] + t + data[y1][x1+1:]

        m = Map()
        m.data = tuple(data)
        return m

    def insert_at(self, row, line):
        data = list(self.data)
        data.insert(row, line)
        self.data = tuple(data)

    def complete(self):
        for y, row in enumerate(self.data):
            for x, e in enumerate(row):
                if e in "ABCD" and x != target[e]:
                    return False
        return True

    def can_move(self, p):
        x, y = p
        if y == 1:
            return True
        if self.data[y-1][x] != ".":
            return False
        e = self.data[y][x]
        if x != target[e]:
            return True
        for yt in range(len(self.data)-1, y, -1):
            t = self.data[yt][x]
            if t == "#":
                pass
            elif t != e:
                return True
        return False

    def find_home(self, e, start):
        lst = []
        x0, y0 = start
        x1 = target[e]
        if x1 > x0:
            dx = 1
        else:
            dx = -1

        # walk horizontally
        for x in range(x0+dx, x1+dx, dx):
            if self.data[y0][x] != ".":
                return lst

        # check vertically
        for y1 in range(len(self.data)-2, 1, -1):
            t = self.data[y1][x1]
            if t == ".":
                lst.append((x1, y1))
                break
            elif t != e:
                break

        return lst

    def find_hallways(self, e, start):
        lst = []
        x0, y0 = start

        # walk up
        for y in range(y0-1, 0, -1):
            if self.data[y][x0] != ".":
                return lst

        # walk left
        for x in range(x0-1, -1, -1):
            if self.data[y][x] != ".":
                break
            elif x in (3, 5, 7, 9):
                pass
            else:
                lst.append((x, y))

        # walk right
        for x in range(x0+1, 13):
            if self.data[y][x] != ".":
                break
            elif x in (3,5,7,9):
                pass
            else:
                lst.append((x, y))

        return lst

    def __hash__(self):
        return hash(self.data)

    def __eq__(self, other):
        return self.data == other.data

    def __lt__(self, other):
        return self.data < other.data

def manhattan(a, b):
    return abs(b[0]-a[0]) + abs(b[1]-a[1])

def decisions(m):
    res = []
    for y, row in enumerate(m.data):
        for x, e in enumerate(row):
            src = (x, y)
            if e in "ABCD" and m.can_move(src):
                if y == 1:
                    dsts = m.find_home(e, src)
                else:
                    dsts = m.find_hallways(e, src)

                for dst in dsts:
                    en = manhattan(src, dst) * energy[e]
                    res.append((en, m.swap(src, dst)))

    return res

def dijkstra(start):
    visited = {}
    heap = []

    visited[start] = 0
    heapq.heappush(heap, (0, start))

    while heap:
        dist, m = heapq.heappop(heap)
        if m.complete():
            return dist

        for dd, mm in decisions(m):
            ndist = dist + dd
            if not mm in visited or ndist < visited[mm]:
                visited[mm] = ndist
                heapq.heappush(heap, (ndist, mm))

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        m = Map()
        m.from_txt(txt)
        print("Part1:", dijkstra(m))

        m.insert_at(3,"  #D#C#B#A#")
        m.insert_at(4,"  #D#B#A#C#")
        print("Part2:", dijkstra(m))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

File: day23/day23.cpp
#include <cassert>
#include <algorithm>
#include <fstream>
#include <list>
#include <queue>
#include <unordered_map>
#include <vector>
#include <fmt/format.h>

using namespace std;

static const unsigned ENERGY[] = { 1, 10, 100, 1000 };

struct Point
{
	int x;
	int y;
};

static vector<Point> points;

struct Map
{
	char data[7][13];
	unsigned energy;
	unsigned priority;

	bool complete() const
	{
		for (int y = 2; y < 7; y++)
		{
			for (int i = 0; i < 4; i++)
			{
				if (data[y][3 + i * 2] == '#')
				{
					return true;
				}
				if (data[y][3 + i * 2] != 'A' + i)
				{
					return false;
				}
			}
		}
		return true;
	}

	bool can_move(Point p) const
	{
		if (p.y == 1)
		{
			return true;
		}
		if (data[p.y-1][p.x] != '.')
		{
			return false;
		}
		int e = data[p.y][p.x];
		if (p.x !=  3 + 2 * (e - 'A'))
		{
			return true;
		}
		for (int y = 5; y > p.y; y--)
		{
			if (data[y][p.x] == '#')
			{
			}
			else if (data[y][p.x] != e)
			{
				return true;
			}
		}
		return false;
	}

	void set_priority()
	{
		unsigned prio = 0;
		for (auto [x, y]: points)
		{
			int e = data[y][x];
			if ('A' <= e && e <= 'D')
			{
				int tx = 3 + 2 * (e - 'A');
				prio += (tx > x ? (tx - x) : (x - tx)) * ENERGY[e-'A'];
			}
		}
		priority = energy + prio;
	}

	void insert_at(int y, string row)
	{
		size_t len = row.size();
		if (len > 13)
		{
			len = 13;
		}
		memmove(data[y+1], data[y], (7 - y- 1) * 13);
		memmove(data[y], row.data(), len);
	}

	void build_points(vector<Point>& points) const
	{
		points.clear();
		for (int y = 0; y < 7; y++)
		{
			for (int x = 0; x < 13; x++)
			{
				if (y == 1 && (x == 3 || x == 5 || x == 7 || x == 9))
				{
					continue;
				}
				switch (data[y][x])
				{
				case 'A':
				case 'B':
				case 'C':
				case 'D':
				case '.':
					points.push_back({x, y});
				default:
					break;
				}
			}
		}
	}

	bool operator==(const Map& other) const
	{
		return !memcmp(data, other.data, sizeof(data));
	}
};

istream& operator>>(istream& in, Map& m)
{
	m.energy = 0;
	memset(m.data, '#', sizeof(m.data));

	int y = 0;
	int x = 0;
	char ch;
	while (in.get(ch) && y < 7)
	{
		switch (ch)
		{
		case '.':
		case 'A':
		case 'B':
		case 'C':
		case 'D':
			if (x < 13)
			{
				m.data[y][x] = ch;
			}
		case ' ':
		case '#': x++; break;
		case '\n': y++; x = 0; break;
		}
	}
	return in;
}

namespace std
{
	template<>
	struct hash<Map>
	{
		size_t operator()(const Map& m) const
		{
			size_t hash = 5381;
			for (auto [x, y]: points)
			{
				hash = (hash * 33) ^ m.data[y][x];
			}
			return hash;
		}
	};
}

struct MapCmp
{
	bool operator()(const Map& a, const Map& b) const
	{
		return a.priority > b.priority;
	}
};

using Dict = unordered_map<Map, unsigned>;
using Heap = priority_queue<Map, vector<Map>, MapCmp>;
using List = list<Map>;

static void find_home(const Map& m, int e, Point p, vector<Point>& dst)
{
	int x = 3 + 2 * (e - 'A');

	// walk left/right
	int dx = p.x < x ? 1 : -1;
	while (p.x != x)
	{
		p.x += dx;
		if (m.data[p.y][p.x] != '.')
		{
			return;
		}
	}

	// check vertically
	for (p.y = 5; p.y >= 2; p.y--)
	{
		int v = m.data[p.y][p.x];
		if (v == '#' || v == e)
		{
			// ignore
		}
		else if (v == '.')
		{
			dst.push_back(p);
			break;
		}
		else
		{
			break;
		}
	}
}

static void find_hallways(const Map& m, int e, Point p, vector<Point>& dst)
{
	// walk up
	for (p.y--; p.y > 1; p.y--)
	{
		if (m.data[p.y][p.x] != '.')
		{
			return;
		}
	}

	// walk left
	int x = p.x;
	for (p.x = x - 1; ; p.x--)
	{
		if (m.data[p.y][p.x] != '.')
		{
			break;
		}
		else if (p.x != 3 && p.x != 5 && p.x != 7 && p.x != 9)
		{
			dst.push_back(p);
		}
	}

	// walk right
	for (p.x = x + 1; ; p.x++)
	{
		if (m.data[p.y][p.x] != '.')
		{
			break;
		}
		else if (p.x != 3 && p.x != 5 && p.x != 7 && p.x != 9)
		{
			dst.push_back(p);
		}
	}
}

static unsigned manhattan(Point a, Point b)
{
	return (a.x > b.x ? (a.x - b.x) : (b.x - a.x))
		+ (a.y > b.y ? (a.y - b.y) : (b.y - a.y));
}

static List decisions(const Map& m)
{
	List res;
	vector<Point> dst;
	for (auto from : points)
	{
		int e = m.data[from.y][from.x];
		if ('A' <= e && e <= 'D' && m.can_move(from))
		{
			dst.clear();
			if (from.y == 1)
			{
				find_home(m, e, from, dst);
			}
			else
			{
				find_hallways(m, e, from, dst);
			}

			for (auto to: dst)
			{
				Map n(m);
				n.data[from.y][from.x] = '.';
				n.data[to.y][to.x] = e;
				n.energy += manhattan(from, to) * ENERGY[e - 'A'];
				res.push_back(move(n));
			}
		}
	}
	return res;
}

static size_t astar(const Map& start)
{
	size_t result = 0;
	Dict visited;
	Heap queue;

	start.build_points(points);
	visited[start] = 0;
	queue.push(start);
	while (!queue.empty())
	{
		Map p = queue.top();
		queue.pop();

		if (p.complete())
		{
			result = p.energy;
			break;
		}

		for (auto &c : decisions(p))
		{
			auto it = visited.find(c);
			if (it == visited.end() || c.energy < it->second)
			{
				c.set_priority();
				visited[c] = c.energy;
				queue.push(c);
			}
		}
	}

	return result;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}
	Map m;
	in >> m;
	in.close();

	fmt::print("Part1: {}\n", astar(m));

	m.insert_at(3, "  #D#C#B#A");
	m.insert_at(4, "  #D#B#A#C");
	fmt::print("Part2: {}\n", astar(m));
	return 0;
}
